lib/vkloader.c

```diff
@@ -146,38 +169,179 @@ ktxVulkanDeviceInfo_Construct(ktxVulkanDeviceInfo* This,
 
+#define LOAD_EXT_FUNC(member, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)ktxLoadVulkanFunction(#fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+#define LOAD_INST_FUNC(member, instance, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)((member).vkGetInstanceProcAddr)((instance), #fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+#define LOAD_DEVICE_FUNC(member, device, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)((member).vkGetDeviceProcAddr)((device), #fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+
+    This->instance = instance;
     This->physicalDevice = physicalDevice;
     This->device = device;
     This->queue = queue;
     This->cmdPool = cmdPool;
     This->pAllocator = pAllocator;
-    vkGetPhysicalDeviceMemoryProperties(physicalDevice,
+
+    ktxVulkanFunctions funcs;
+    memset(&funcs, 0, sizeof(ktxVulkanFunctions));
+
+    if (pFunctions != NULL)
+        funcs = *pFunctions;
+
+    if (instance == VK_NULL_HANDLE || pFunctions == NULL) {
+        // This is the old behavior, where no functions where specified. We take a shot at loading them dynamically.
+#if defined(KTX_USE_FUNCPTRS_FOR_VULKAN)
+        // Delay loading not supported so must do it ourselves.
+        if (!ktxVulkanModuleHandle) {
+            ktx_error_code_e kresult = ktxLoadVulkanLibrary();
+            if (kresult != KTX_SUCCESS)
+                return kresult;
+        }
+#endif
+
+        // if pFunctions are null, then we try to load the minimum number of required functions.
+        if (pFunctions == NULL) {
+            LOAD_EXT_FUNC(funcs, vkGetInstanceProcAddr);
+        }
+
+        // If we have no instance, we will need to bind the instance instructions and the function instructions
+        if (instance == VK_NULL_HANDLE) {
+            LOAD_EXT_FUNC(funcs, vkGetDeviceProcAddr);
+
+            // These require an instance and we don't have that.
+            LOAD_EXT_FUNC(funcs, vkGetPhysicalDeviceFormatProperties);
+            LOAD_EXT_FUNC(funcs, vkGetPhysicalDeviceImageFormatProperties);
+            LOAD_EXT_FUNC(funcs, vkGetPhysicalDeviceMemoryProperties);
+        }
+    }
+
+    // We now load the rest of the functions in.
+    LOAD_INST_FUNC(funcs, instance, vkGetPhysicalDeviceFormatProperties);
+    LOAD_INST_FUNC(funcs, instance, vkGetPhysicalDeviceMemoryProperties);
+    LOAD_INST_FUNC(funcs, instance, vkGetPhysicalDeviceImageFormatProperties);
+
+
+    // We now get to device functions.
+    LOAD_INST_FUNC(funcs, instance, vkGetDeviceProcAddr);
```

Comment:
This and the 3 calls above should be in an else to avoid trying to load them when there is not instance.

Reply:
They don't have to be in an else. The `LOAD_INST_FUNC` checks if the member is initialized before trying a load. If we have no instance, we load them above with the "old" global functions and so they won't be loaded here.

Reply:
Got it. Okay.